pkg/driver/aws-ebs/aws_ebs_tags_controller.go

```diff
@@ -0,0 +1,446 @@
+package aws_ebs
+
+import (
+	"context"
+	"fmt"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"os"
+	"reflect"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorapi "github.com/openshift/api/operator/v1"
+	"github.com/openshift/csi-operator/pkg/clients"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+const (
+	awsSecretNamespace     = "openshift-cluster-csi-drivers"
+	awsSecretName          = "ebs-cloud-credentials"
+	infrastructureResource = "cluster"
+	driverName             = "ebs.csi.aws.com"
+)
+
+type EBSVolumeTagsController struct {
+	name          string
+	commonClient  *clients.Clients
+	eventRecorder events.Recorder
+	failedQueue   workqueue.RateLimitingInterface
+}
+
+func NewEBSVolumeTagsController(
+	ctx context.Context,
+	name string,
+	commonClient *clients.Clients,
+	eventRecorder events.Recorder) factory.Controller {
+
+	c := &EBSVolumeTagsController{
+		name:          name,
+		commonClient:  commonClient,
+		eventRecorder: eventRecorder,
+		failedQueue:   workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+	// Define the informer for watching Infrastructure resource changes
+	infraInformer := c.commonClient.ConfigInformers.Config().V1().Infrastructures().Informer()
+
+	// Add event handler to process updates only when ResourceTags change
+	infraInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
We don't have do this. If we are using `factory.WithInformers`, we will reconcile this within `Sync` function.

Reply:
Hey @gnufied I was bit confused with this!! basically I don't want to run reconciliation on every change on InfraStructure resource. I need to run reconciliation only if there is a any change in `infra.Status.PlatformStatus.AWS.ResourceTags`.

Can you suggest a better way to do this so that we can remove the unnecessarily computes.

Reply:
> How will this work when controller is restarted? You are also doing `WithInformers` below and hence any change in `infra` object will still trigger `Sync`.

What I am thinking is on every restart I would like to run a sync function and update the volumes if there is a change and further On each change in resource-tags we would again run the sync.


> So, what we have currently is worst of both the worlds. If I were to design this, I will probably make a hash of sorted tags and annotate PV with tag hash. If tag hash annotation in PV and computed tag hash don't change, then I will not update the PV or else I will.

This looks easy way to manage this!! I will then add a new field in controller struct which will have the updated hash for the sorted map of tags! now in each reconciliation I will update tags only if the hash is different from the other one? Does this sounds better to you??

Reply:
Added Thanks For Suggestion!!